### ISSUE TITLE
A subclassed generic type is not detected and raises a mypy warning

### ISSUE THREAD

**Issue by exhuma** [python/mypy#7322]
The way I understand covariance (and I am not certain I understood it correctly), the code below should pass the type-check as the type `Container[SpecialisedItemType]` is a subtype of `Container[ItemType]` assuming they are treated as covariant.

Am I doing something wrong here or is this a limitation of mypy? Or, is there a way that I can tell mypy that these types are covariant?

The following code triggers a mypy error (at the indicated location) but I would expect it not to. The error it reports is:

    generic_covariance.py:28: error: Return type "Container[SpecialisedItemType]" of "some_function" incompatible with return type "Container[ItemType]" in supertype "Interface"

The minimal code to reproduce this:

```
from typing import Generic, TypeVar

T = TypeVar('T', bound='ItemType')


class Container(Generic[T]):
    pass


class ItemType:
    pass


class SpecialisedItemType(ItemType):
    pass


class Interface:

    def some_function(self) -> Container[ItemType]:
        pass


class ConcreteClass(Interface):

    def some_function(self) -> Container[SpecialisedItemType]:  # <-- Error reported here
        pass
```

**Comment by ilevkivskyi** [comment]
Variance must be declared in mypy, please read the docs https://mypy.readthedocs.io/en/latest/generics.html#variance-of-generic-types
